### ISSUE TITLE
Port twisted.web.template to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7811]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#7811|
|Type|enhancement|
|Created|2015-03-15 08:09:20Z|
|Branch|https://github.com/twisted/twisted/tree/twwebtemplate-py3-7811-4|

`twisted.web.template` should work on Python 3

<details><summary>Searchable metadata</summary>

```
trac-id__7811 7811
type__enhancement enhancement
reporter__hawkowl hawkowl
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_twwebtemplate_py3_7811_4 branches/twwebtemplate-py3-7811-4
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1426406960442982 1426406960442982
changetime__1436254582488383 1436254582488383
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [44040]) Branching to twwebtemplate-py3-7811.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

(In [45120]) Branching to twwebtemplate-py3-7811-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

(In [45125]) Branching to twwebtemplate-py3-7811-3.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl commented|
|-|-|

Porting done, builders spun, etc etc.

This does have the change of making the attributes sorted always, which sidesteps a PyPy/Python 3 bug where it would get out of order and fail because of string comparison.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Great work. Thanks!

Maybe we  can move OrderedDict into twisted.python.compat and always import it from there.

Why do we need this change
```
if data and data[-1:] == b'-':
```

the `data[-1]` to `data[-1:]` part?


--------

Just minor comments. Feel free to review my comments and merge.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="SamyCookie's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|SamyCookie commented|
|-|-|

Is Python 2.6 still a target ? If not, 'OrderedDict' is available with Python 2.7 and all compatibility stuff for it can be removed.

https://docs.python.org/2/library/collections.html#collections.OrderedDict

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl commented|
|-|-|

(In [45184]) Branching to twwebtemplate-py3-7811-4.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl commented|
|-|-|

I re-enabled the tests which t.w.t now allow to run. As such, putting up for review again.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Changes look good.

It looks like this ticket also fixe t.web.utils ([#6178](https://github.com/twisted/twisted/issues/6178)) , beside t.web.templates. Maybe the newsfile should be updated and the associated ticket closed as duplicate.

I think that the first part of `stringToCheckFor` could be improved to reduce duplication.

Please address my comments and merge.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl commented|
|-|-|

Newsfile amended.

I had a play around with stringToCheckFor and couldn't come up with anything cleaner.

Thanks for the review. :)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_12"></a>|@hawkowl set status to `closed`|
|-|-|

(In [45212]) Merge twwebtemplate-py3-7811-4: Port twisted.web.template to Python 3

Author: hawkowl
Reviewer: adiroiban
Fixes: [#7811](https://github.com/twisted/twisted/issues/7811)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_13"></a>|@hawkowl commented|
|-|-|

(In [45217]) Merge twwebtemplate-py3-7811-4: Port twisted.web.template to Python 3

Author: hawkowl
Reviewer: adiroiban
Fixes: [#7811](https://github.com/twisted/twisted/issues/7811)
